File: Problem876/solution.cpp
/* 
Let n be the number of nodes in the list

Time Complexity: O(n)
Space Complexity: O(1)

A simple and easy two pointers solution.

*/

#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
 
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode* p = head;
        ListNode* q = head;
        while(p->next != nullptr){
            if(p->next->next != nullptr){
                p = p->next->next;
            }else{
                p = p->next;
            }
            q = q->next;
        }
        return q;
    }
};

int main(){

    Solution sol = Solution();
    ListNode* p = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5, new ListNode(6))))));

    cout << sol.middleNode(p)->val << endl;

    delete p->next;
    delete p;

}

File: Problem876/solution.js
/* 
Let n be s.length

Time Complexity: O(n)
Space Complexity: O(1)

A solution using two pointers algorithm.
*/

/**
 * @param {string} s
 * @return {number}
 */
var minimumLength = function(s) {
    let p = 0;
    let q = s.length-1;
    for(let i = 0; i < s.length; i++){
        if(s[p] == s[q] && q > p){
            let letter = s[p]
            p += 1;
            q -= 1;
            while((s[p] == letter || s[q] == letter) && q > p){
                if(s[p] == letter){p += 1;}
                if(s[q] == letter){q -= 1;}
                if(q == p && s[p] == letter){p += 1;}
            }
        }else{break;}
    }
    return q - p + 1;
};

let str = 'aabccabba'
console.log(minimumLength(str))